src/v/raft/replicate_batcher.cc

```diff
@@ -33,13 +33,25 @@ replicate_batcher::replicate_batcher(consensus* ptr, size_t cache_size)
   : _ptr(ptr)
   , _max_batch_size_sem(cache_size) {}
 
-ss::future<result<replicate_result>> replicate_batcher::replicate(
+replicate_stages replicate_batcher::replicate(
   std::optional<model::term_id> expected_term, model::record_batch_reader&& r) {
-    return do_cache(expected_term, std::move(r)).then([this](item_ptr i) {
-        return _lock.with([this] { return flush(); }).then([i] {
-            return i->_promise.get_future();
-        });
-    });
+    ss::promise<> enqueued;
+    auto enq_f = enqueued.get_future();
+    auto f = ss::with_gate(
+      _ptr->_bg,
+      [this,
+       r = std::move(r),
+       expected_term,
+       enqueued = std::move(enqueued)]() mutable {
+          return do_cache(expected_term, std::move(r))
```

Comment:
if `do_cache` fails we never set `enqueued` promise and return (unresolved, failed) replicate_stages pair. It may lead to a situation when a caller will waiting on `enq_f`. it seem like a problem, what do you think?

Reply:
Great suggestion I will add exception forwarding.